### PR TITLE
Refactor Hdfs #300

### PR DESCRIPTION
**Pull request by andrewsmartin**
@ravwojdyla @nevillelyh could you please review?

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -295,7 +295,7 @@ lazy val scioHdfs: Project = Project(
   )
 ).dependsOn(
   scioCore,
-  scioTest % "test",
+  scioTest % "test->test",
```
Comment (nevillelyh) [inline build.sbt]: what does `"test->test"` mean?

File: scio-hdfs/src/main/scala/com/spotify/scio/hdfs/HdfsFileStorage.scala [hunk 2]
```diff
@@ -0,0 +1,65 @@
+package com.spotify.scio.hdfs
```
Comment (nevillelyh) [inline scio-hdfs/src/main/scala/com/spotify/scio/hdfs/HdfsFileStorage.scala]: copyright header

File: scio-hdfs/src/main/scala/com/spotify/scio/hdfs/HdfsFileStorage.scala [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+      !path.getName.startsWith("_") && !path.getName.startsWith(".")
+  }
+
+  override protected def listFiles: Seq[Path] = {
+    val conf = new Configuration()
+    val fs = FileSystem.get(new URI(path), conf)
+
+    fs
+      .listStatus(new org.apache.hadoop.fs.Path(path), pathFilter)
+      .map(status => new File(path + "/" + status.getPath.getName).toPath).toSeq
```
Comment (nevillelyh) [inline scio-hdfs/src/main/scala/com/spotify/scio/hdfs/HdfsFileStorage.scala]: there's a better way than string concat right? `status.getPath` doesn't give you the full path?

File: scio-hdfs/src/main/scala/com/spotify/scio/hdfs/HdfsFileStorage.scala [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  override protected def getObjectInputStream(path: Path): InputStream = {
+    val hPath = new org.apache.hadoop.fs.Path(path.toString)
+    val conf = new Configuration()
+    val factory = new CompressionCodecFactory(conf)
+    val fs = FileSystem.get(path.toUri, conf)
+    val codec = factory.getCodec(hPath)
```
Comment (nevillelyh) [inline scio-hdfs/src/main/scala/com/spotify/scio/hdfs/HdfsFileStorage.scala]: Does this work for formats like avro?

Comment (nevillelyh) [inline scio-hdfs/src/main/scala/com/spotify/scio/hdfs/HdfsFileStorage.scala]: Nevermind it's copied from package.scala so should be fine.

File: build.sbt (hunk 1 above)
Comment (andrewsmartin) [inline build.sbt]: From http://www.scala-sbt.org/0.12.2/docs/Getting-Started/Multi-Project.html:

"A useful declaration is "test->test" which means Test depends on Test. This allows you to put utility code for testing in bar/src/test/scala and then use that code in foo/src/test/scala, for example."

Comment (nevillelyh) [inline build.sbt]: Great!

Comment (andrewsmartin) [inline build.sbt]: "test" by itself is short for "test->compile"

File: scio-hdfs/src/main/scala/com/spotify/scio/hdfs/HdfsFileStorage.scala (hunk 2 above)
Comment (andrewsmartin) [inline scio-hdfs/src/main/scala/com/spotify/scio/hdfs/HdfsFileStorage.scala]: Added

**Comment by nevillelyh** [comment]
:+1:

File: scio-hdfs/src/main/scala/com/spotify/scio/hdfs/HdfsFileStorage.scala [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    if (codec != null) {
+      codec.createInputStream(fs.open(hPath))
+    } else {
+      fs.open(hPath)
+    }
+  }
+
+  override protected def getAvroSeekableInput(path: Path): SeekableInput =
+    new SeekableInput {
```
Comment (ravwojdyla) [inline scio-hdfs/src/main/scala/com/spotify/scio/hdfs/HdfsFileStorage.scala]: There is no support for this in Avro already?

File: scio-hdfs/src/main/scala/com/spotify/scio/hdfs/HdfsFileStorage.scala [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+    fs
+      .listStatus(new org.apache.hadoop.fs.Path(path), pathFilter)
+      .map(status => new File(status.getPath.toString).toPath).toSeq
+  }
+
+  override protected def getObjectInputStream(path: Path): InputStream = {
+    val hPath = new org.apache.hadoop.fs.Path(path.toString)
+    val conf = new Configuration()
+    val factory = new CompressionCodecFactory(conf)
```
Comment (ravwojdyla) [inline scio-hdfs/src/main/scala/com/spotify/scio/hdfs/HdfsFileStorage.scala]: I have a vague memory of something like this being already supported in avro itself?

File: scio-test/src/test/scala/com/spotify/scio/io/TapTest.scala [hunk 7]
```diff
@@ -33,7 +33,34 @@ import org.apache.commons.io.FileUtils
...
+  def runWithFileFuture[T](fn: ScioContext => Future[Tap[T]]): Tap[T] =
+    runWithFuture(ScioContext())(fn)
+
+  def runWithFuture[T](sc: ScioContext)(fn: ScioContext => Future[Tap[T]]): Tap[T] = {
+    val f = fn(sc)
+    sc.close()
+    f.waitForResult()
+  }
+
+  def tmpDir: File = new File(
```
Comment (ravwojdyla) [inline scio-test/src/test/scala/com/spotify/scio/io/TapTest.scala]: There is `Files.createTempDirectory` etc already

File: build.sbt (hunk 1 above)
Comment (ravwojdyla) [inline build.sbt]: 👍

**Pull request merged**